In [62]:
import pandas as pd
import numpy as np
import pandas_profiling as prof
from decimal import Decimal
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
 
data = pd.read_csv("/Users/ksomes/Downloads/listings.csv") 

In [42]:
#drop redundant info and fields not useful for analysis
sub=data.drop(['id','listing_url','scrape_id','last_scraped','summary','space','description','experiences_offered'
              , 'neighborhood_overview', 'notes', 'transit', 'access', 'interaction', 'house_rules',
              'thumbnail_url', 'medium_url', 'picture_url', 'xl_picture_url', 'host_url', 'host_thumbnail_url',
              'host_picture_url', 'country_code', 'country','amenities', 'minimum_minimum_nights',
              'maximum_minimum_nights','minimum_maximum_nights', 'maximum_maximum_nights','minimum_nights_avg_ntm',
              'maximum_nights_avg_ntm', 'availability_30', 'availability_365','availability_90','has_availability',
               'calculated_host_listings_count','calculated_host_listings_count_shared_rooms',
               'is_business_travel_ready','host_about', 'host_acceptance_rate', 'host_total_listings_count',
              'jurisdiction_names','license','monthly_price','square_feet','weekly_price', 'requires_license'], axis=1)
print(list(sub.columns))

['name', 'host_id', 'host_name', 'host_since', 'host_location', 'host_response_time', 'host_response_rate', 'host_is_superhost', 'host_neighbourhood', 'host_listings_count', 'host_verifications', 'host_has_profile_pic', 'host_identity_verified', 'street', 'neighbourhood', 'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'city', 'state', 'zipcode', 'market', 'smart_location', 'latitude', 'longitude', 'is_location_exact', 'property_type', 'room_type', 'accommodates', 'bathrooms', 'bedrooms', 'beds', 'bed_type', 'price', 'security_deposit', 'cleaning_fee', 'guests_included', 'extra_people', 'minimum_nights', 'maximum_nights', 'calendar_updated', 'availability_60', 'calendar_last_scraped', 'number_of_reviews', 'number_of_reviews_ltm', 'first_review', 'last_review', 'review_scores_rating', 'review_scores_accuracy', 'review_scores_cleanliness', 'review_scores_checkin', 'review_scores_communication', 'review_scores_location', 'review_scores_value', 'instant_bookable', 'cancellation_p

In [96]:
def money_to_decimal(x):
    x = x.replace("$", "").replace(",", "").replace(" ", "")
    return float(x)
def rem_percent(x):
    x=x.replace("%","")
    return float(x)/100
def truncate(n):
    return int(n * 1000) / 1000

In [68]:
#converts objects with money values into decimal values to become continous attribute
sub.cleaning_fee = sub.cleaning_fee.astype(str)
sub.extra_people = sub.extra_people.astype(str)
sub.security_deposit = sub.security_deposit.astype(str)
sub.price = sub.price.astype(str)
sub.loc[:,'price'] = sub.loc[:,'price'].apply(money_to_decimal)
sub.loc[:,'cleaning_fee'] = sub.loc[:,'cleaning_fee'].apply(money_to_decimal)
sub.loc[:,'extra_people'] = sub.loc[:,'extra_people'].apply(money_to_decimal)
sub.loc[:,'security_deposit'] = sub.loc[:,'security_deposit'].apply(money_to_decimal)

#imputations
sub['price']=sub.price.mask(sub.price == 0,sub.price.median())
sub.cleaning_fee=sub.cleaning_fee.fillna(sub.cleaning_fee.median())
sub.first_review=sub.first_review.fillna('2019-08-01')
sub['first_review'] =  pd.to_datetime(sub['first_review'],
                              format='%Y-%m-%d')
sub.host_response_rate = sub.host_response_rate.astype(str)
sub.loc[:,'host_response_rate'] = sub.loc[:, 'host_response_rate'].apply(rem_percent)
sub.host_response_rate=sub.host_response_rate.fillna(sub.host_response_rate.median())
sub['host_since'] =  pd.to_datetime(sub['host_since'],
                              format='%Y-%m-%d')
sub.last_review=sub.last_review.fillna('2019-08-01')
sub['last_review'] =  pd.to_datetime(sub['last_review'],
                              format='%Y-%m-%d')
sub.review_scores_accuracy=sub.review_scores_accuracy.fillna(truncate(sub.review_scores_accuracy.median()))
sub.review_scores_checkin=sub.review_scores_checkin.fillna(truncate(sub.review_scores_checkin.median()))
sub.review_scores_cleanliness=sub.review_scores_cleanliness.fillna(truncate(sub.review_scores_cleanliness.median()))
sub.review_scores_communication=sub.review_scores_communication.fillna(truncate(sub.review_scores_communication.median()))
sub.review_scores_location=sub.review_scores_location.fillna(truncate(sub.review_scores_location.median()))
sub.review_scores_rating=sub.review_scores_rating.fillna(truncate(sub.review_scores_rating.median()))
sub.review_scores_value=sub.review_scores_value.fillna(truncate(sub.review_scores_value.median()))
sub.reviews_per_month=sub.reviews_per_month.fillna(sub.reviews_per_month.median())
sub.security_deposit=sub.security_deposit.fillna(sub.security_deposit.median())

In [103]:
sub.security_deposit=sub.security_deposit.fillna(sub.security_deposit.median())
print(sub.security_deposit)

0         350.0
1         200.0
2         500.0
3         100.0
4         300.0
5         450.0
6         120.0
7         120.0
8         500.0
9           0.0
10        450.0
11        200.0
12        200.0
13        500.0
14        200.0
15        200.0
16          0.0
17        500.0
18        120.0
19        300.0
20          0.0
21        120.0
22        300.0
23          0.0
24        300.0
25        120.0
26        120.0
27          0.0
28        150.0
29        120.0
          ...  
48834     500.0
48835       0.0
48836     120.0
48837     120.0
48838       0.0
48839     120.0
48840     350.0
48841     500.0
48842       0.0
48843     250.0
48844     120.0
48845     120.0
48846     150.0
48847     120.0
48848       0.0
48849       0.0
48850    1000.0
48851       0.0
48852       0.0
48853    1000.0
48854     120.0
48855       0.0
48856     100.0
48857       0.0
48858       0.0
48859     120.0
48860     120.0
48861       0.0
48862     300.0
48863     120.0
Name: security_deposit, 

0         225.0
1         150.0
2          89.0
3          80.0
4         200.0
5          60.0
6          79.0
7          79.0
8         116.0
9         150.0
10        135.0
11         85.0
12         89.0
13        120.0
14        140.0
15        215.0
16        140.0
17         99.0
18        190.0
19        299.0
20        115.0
21        110.0
22        120.0
23         60.0
24         80.0
25        150.0
26        180.0
27         50.0
28         52.0
29         55.0
          ...  
48834     499.0
48835     180.0
48836     105.0
48837     150.0
48838     180.0
48839      65.0
48840     315.0
48841    1295.0
48842     250.0
48843     895.0
48844     150.0
48845     280.0
48846     395.0
48847     350.0
48848     100.0
48849     180.0
48850     190.0
48851     236.0
48852     180.0
48853     300.0
48854     709.0
48855     180.0
48856     135.0
48857     300.0
48858     130.0
48859      65.0
48860      60.0
48861     239.0
48862      75.0
48863     239.0
Name: price, Length: 488

In [65]:
profile = sub.profile_report(title="AirBnb Profiling", correlation_overrides=["recclass"])
profile.to_file(output_file=("/Users/ksomes/Downloads/airbnb_profile.html"))

## Data description
__accomodates:__ The maximum number of people the rental can accomodate *(numeric)*  
__availability_60:__ Number of days the space is available within the next 60 days *(numeric)*  
__bathrooms:__ The number of bathrooms in the rental *(numeric)*  
__bed_type:__ The type of bed available *(categorical)*  
__bedrooms:__ The number of separate bedrooms in the rental *(numeric)*  
__beds:__ The number of beds in the rental *(numeric)*  
__calculated_host_listings_count_entire_homes:__ Number of listings a host has where an entire home is offered *(numeric)*  
__calculated_host_listings_count_private_rooms:__ Number of listings a host has with private rooms offered *(numeric)*  
__calendar_last_scraped:__ Date the calendar information was last pulled *(date)*  
__calendar_updated:__ Last time the host updated their calendar information *(date)*  
__cancellation_policy:__ Ranking of how strict the host is on their cancellation policy *(categorical)*  
__city:__ The city/borough the listing is located in *(categorical)*  
__cleaning_fee:__ The surcharge for cleaning the listing *(numeric)*  
__extra_people:__ The surcharge for each additional person staying in the rental *(numeric)*  
__first_review:__ The date of the first review of the listing *(date)*  
__guests_included:__ The number of guests included in the price *(numeric)*  
__host_has_profile_pic:__ Indicator of presence of profile picture on site *(Boolean)*  
__host_id:__ Host identifier *(numeric)*  
__host_identity_verified:__ Indicator of host's identity verification using offline source *(Boolean)*  
__host_is_superhost:__ Indicator of super host status (exemplary host behavior) *(Boolean)*  
__host_listings_count:__ The number of listings the host has *(numeric)*  
__host_location:__ The city the host is located in *(categorical)*  
__host_name:__ Name of host *(categorical)*  
__host_neighborhood:__ The neighborhood the host is located in *(categorical)*  
__host_response_rate:__ The rate which a host responds to contact from lessor *(numeric)*  
__host_response_time:__ Bucket of how long the host takes to respond *(categorical)*  
__host_since:__ The date of the first hosting *(date)*  
__host_verifications:__ Ways in which the host was verified *(categorical)*  
__instant_bookable:__ Indicator on ability to book instantly *(Boolean)*  
__is_location_exact:__ Indicator on if location is exact *(Boolean)*  
__last_review:__ Date the last review of the listing occured *(date)*  
__latitude:__ Latitude of the listing *(numeric)*  
__longitude:__ Longitude of the listing *(numeric)*  
__market:__ The market the listing is in (i.e., Adirondacks, NYC) *(categorical)*  
__maximum_nights:__ The maximum number of consecutive days the listing is available *(numeric)*  
__minimum_nights:__ The minimum number of days the place can be rented *(numeric)*  
__name:__ Given name of the listing *(categorical)*  
__neighborhood:__ The neighborhood the listing is located in *(categorical)*  
__number_of_reviews__: The number of reviews on a listing *(numeric)*  
__number_of_reviews_ltm:__ The number of reviews on a long term listing *(numeric)*  
__price:__ The nightly price of a listing *(numeric)*  
__property_type:__ The residence type of the listing *(numeric)*  
__require_guest_phone_verification:__ Indicator on whether phone verification of guest is requied *(Boolean)*  
__require_guest_profile_picture:__  Indicator on whether host requires guest to post profile picture *(Boolean)*  
__review_scores_accuracy:__ The average rating of all guests on the accuracy of the listing's description *(numeric)*  
__review_scores_checkin:__ The average rating of guests' review of the check in process *(numeric)*  
__review_scores_cleanliness:__ The average rating of guests' review of cleanliness *(numeric)*  
__review_scores_communication:__ The average rating of guests' review on host communication *(numeric)*  
__review_scores_location:__ The average rating of guests' review of listing location *(numeric)*  
__review_scores_rating:__  The average rating of guests' review of overall host rating *(numeric)*  
__review_scores_value:__ The average rating of guests' review of listing value *(numeric)*  
__reviews_per_month:__ The number of reviews a listing gets per month *(numeric)*  
__room_type:__ The type of space that is available for rent *(categorical)*  
__security_deposit:__ The value of the security deposit required before rent *(numeric)*  
__smart_location:__ The city of the listing - cleaned *(categorical)*    
__state:__  The state the listing is in *(categorical)*   
__street:__ Another variable for city *(categorical)*   
__zipcode:__ The zipcode the listing is in *(numeric)* 


### There are potential outliers with the following variables:
 - maximum nights (placeholders of random values?)
 - minimum nights (same thing?)


### Skewed variables:
 - price
 - number of reviews
 - Continue this!!
 
### Transformed variables


### Duplicate data:
 - Several minimum/maximum night variables conveying the same information
 - Availability variables were highly correlated
 - Calculated host listing variables contained same information
 - Weekly and monthly price were highly correlated with price
 - Country and country code contained same info (redundant for NY listings)
 - Requires_license is exlusively true except for 1 value
 
### Missing values:
Many attributes have thousands of missing values. This is due to the nature of AirBnb's site - reviewers are not required to provide feedback on every category, and hosts do not have to provide information in every single field. Many of the fields available on the site do not apply to every single listing, meaning that often these fields are left blank and it is not a result of error. Certain fields had a high frequency of missing values (e.g. square feet had over 90% missing values), in which we determined not to consider those variables as eligible for analysis. Out of the variables worth considering, the highest percentage of missing values is less than 40%, where we chose to impute the values by using the median of the available values to reduce changes to the shape of the distribution. For categorical variables, the values were not imputed, as "NaN" can serve as it's own factor level.

A key target variable for prediction is "price" which has 10 zero values which will be considered missing and included in imputation, since price is required for listings. 

Host neighborhood was not considered for imputation, since there are many variables that contain similar information that can be used for analysis.

First review and last review were imputed with the date 08/01/2019 with the assumption that dates farther in the past provide information about host experience, which is information we did not want to provide for missing values.

## Descriptive Statistics of Key Variables 
__Price:__  
As shown in the table below, the prices are skewed right (mean is greater than the median), meaning there are more listings with a price below the average than above. 75% of the prices per night are /$175 or less, but there are high-end options up to /$10,000 a night. The prices are widely spread, as shown by the large range (99,990) and standard deviation (236).

In [69]:
#price
sub.price.describe()
sub.price.std()

count    48864.000000
mean       151.474664
std        236.576536
min         10.000000
25%         69.000000
50%        105.000000
75%        175.000000
max      10000.000000
Name: price, dtype: float64

236.57653562777037

__Beds:__  
75% of the listings have 2 beds or less, with a greater concentration of listings having only one bed available. However, there are 5 listings with over 20 times as many beds as the median number of beds, indicating that this variable is highly skewed. The maximum number of beds a listing has is 40--which is much greater than the average number of 1.55. An early hypothesis is that these values with larger values are in specific markets; the market variable will likely have a relationship with the number of beds.

In [70]:
sub.beds.describe()
sub.beds.std()

count    48822.000000
mean         1.548073
std          1.112344
min          0.000000
25%          1.000000
50%          1.000000
75%          2.000000
max         40.000000
Name: beds, dtype: float64

1.1123440223779768